New differential functionality

samjrholt opened the pull request.

lang-m commented on discretisedfield/field.py:
```diff
@@ -2125,7 +2142,7 @@ def div(self):
 
         return sum(
             [
-                getattr(self, self.vdims[i]).derivative(dfu.raxesdict[i])
+                getattr(self, self.vdims[i]).diff(dfu.raxesdict[i])
                 for i in range(self.nvdim)
             ]
         )
```
The extra square brackets can be removed.

lang-m commented on discretisedfield/field.py:
```diff
@@ -2055,7 +2072,7 @@ def grad(self):
+        return self.diff("x") << self.diff("y") << self.diff("z")
 
     @property
     def div(self):
```
The double backslashes in the docstring have to be changed to single backslashes (because it is a raw string). Probably also in the other docstrings.

lang-m commented on discretisedfield/field.py:
```diff
@@ -2125,7 +2142,7 @@ def div(self):
         return sum(
             [
-                getattr(self, self.vdims[i]).derivative(dfu.raxesdict[i])
+                getattr(self, self.vdims[i]).diff(dfu.raxesdict[i])
                 for i in range(self.nvdim)
```
```suggestion
                getattr(self, vdim).diff(vdim) for vdim in self.nvdim
```

lang-m commented on discretisedfield/tests/test_field.py:
```diff
@@ -1218,16 +1228,63 @@ def value_fun(point):
+        def value_fun(point):
+            return point[0] * point[1] * point[2]
+
+        # No Neumann
```
No Dirichlet?

lang-m commented on discretisedfield/tests/test_field.py:
```diff
@@ -1218,16 +1228,63 @@ def value_fun(point):
+        assert np.isclose(f1.diff("y")((1, 7, 1)), 1)
+        assert np.isclose(f1.diff("z")((1, 1, 5)), 1)
+
+        # Neumann
```
Dirichlet?

lang-m reviewed (changes requested):
The changes in the code look good. I commented on a few minor things.

In the tests we are using `np.isclose` and `np.allclose` inconsistenly. It propably does not matter. Is there any particular reason for doing so.

I did not check all the tests in detail (i.e. compute the derivatives to check if the tests are actually correct). I assume you have checked it?

samjrholt commented on discretisedfield/field.py:
```diff
@@ -1915,66 +1916,82 @@ def derivative(self, direction, n=1):
-            pad_width = {dfu.raxesdict[direction]: (1, 1)}
-            padding_mode = "edge"
+            pad_width = {direction: (1, 1)}
+            padding_mode = "symmetric"
```
You are correct, they are equivalent for a single cell but not for more.
`edge` : `abcde|eee`
`symmetric` : `abcde|edc`

samjrholt commented:
> The changes in the code look good. I commented on a few minor things.
> 
> In the tests we are using `np.isclose` and `np.allclose` inconsistenly. It propably does not matter. Is there any particular reason for doing so.
> 
> I did not check all the tests in detail (i.e. compute the derivatives to check if the tests are actually correct). I assume you have checked it?

I have gone through and changed `np.isclose` to `np.allclose` for consistency.
I have also checked the values for the tests given the correct boundary conditions. (And now added those that were missing)

lang-m reviewed (approved):
Looks good to me. You can merge.
